File: day7/part1/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROWS 256
#define MAX_COLS 256

int main(void) {
  FILE *file = fopen("list.txt", "r");
  if (!file) {
    perror("Failed to open list.txt");
    return 1;
  }

  char grid[MAX_ROWS][MAX_COLS];
  int rows = 0;
  int cols = 0;

  while (fgets(grid[rows], MAX_COLS, file)) {

    int len = strlen(grid[rows]);
    if (len > 0 && grid[rows][len - 1] == '\n') {
      grid[rows][len - 1] = '\0';
      len--;
    }
    if (rows == 0) {
      cols = len;
    }
    rows++;
  }
  fclose(file);

  int start_col = -1;
  for (int c = 0; c < cols; c++) {
    if (grid[0][c] == 'S') {
      start_col = c;
      break;
    }
  }

  if (start_col == -1) {
    fprintf(stderr, "Could not find starting position 'S'\n");
    return 1;
  }

  bool *beams = calloc(cols, sizeof(bool));
  bool *next_beams = calloc(cols, sizeof(bool));

  beams[start_col] = true;

  int total_splits = 0;

  for (int r = 1; r < rows; r++) {

    memset(next_beams, 0, cols * sizeof(bool));

    for (int c = 0; c < cols; c++) {
      if (beams[c]) {

        if (grid[r][c] == '^') {

          total_splits++;

          if (c - 1 >= 0) {
            next_beams[c - 1] = true;
          }
          if (c + 1 < cols) {
            next_beams[c + 1] = true;
          }
        } else {

          next_beams[c] = true;
        }
      }
    }

    bool *temp = beams;
    beams = next_beams;
    next_beams = temp;
  }

  printf("Total beam splits: %d\n", total_splits);

  free(beams);
  free(next_beams);

  return 0;
}

File: day7/part1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	MaxRows = 256
	MaxCols = 256
)

func main() {
	file, err := os.Open("../list.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open list.txt: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	grid := make([]string, 0, MaxRows)
	rows := 0
	cols := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if rows == 0 {
			cols = len(line)
		}
		grid = append(grid, line)
		rows++
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	startCol := -1
	for c := 0; c < cols; c++ {
		if grid[0][c] == 'S' {
			startCol = c
			break
		}
	}

	if startCol == -1 {
		fmt.Fprintf(os.Stderr, "Could not find starting position 'S'\n")
		os.Exit(1)
	}

	beams := make([]bool, cols)
	nextBeams := make([]bool, cols)

	beams[startCol] = true

	totalSplits := 0

	for r := 1; r < rows; r++ {
		for i := range nextBeams {
			nextBeams[i] = false
		}

		for c := 0; c < cols; c++ {
			if beams[c] {
				if grid[r][c] == '^' {
					totalSplits++

					if c-1 >= 0 {
						nextBeams[c-1] = true
					}
					if c+1 < cols {
						nextBeams[c+1] = true
					}
				} else {
					nextBeams[c] = true
				}
			}
		}

		beams, nextBeams = nextBeams, beams
	}

	fmt.Printf("Total beam splits: %d\n", totalSplits)
}
